godot-ffi/src/global.rs

```diff
@@ -57,72 +62,30 @@ impl<T> Global<T> {
-                return Err(GlobalLockError::WouldBlock);
-            }
-            Err(TryLockError::Poisoned(poisoned)) => {
-                return Err(GlobalLockError::Poisoned {
-                    // We can likely use `new_unchecked` here, but verifying that it's safe would need somewhat tricky reasoning.
-                    // Since this error condition isn't very common, it is likely not very important to optimize access to the value here.
-                    // Especially since most users will likely not want to access it anyway.
-                    circumvent: GlobalGuard::new(poisoned.into_inner())
-                        .expect("Poisoned global guard should always be initialized"),
-                });
-            }
-        };
-
-        guard.ok_or(GlobalLockError::InitFailed)
-    }
-
-    fn ensure_init(
-        mut mutex_guard: MutexGuard<'_, InitState<T>>,
-        may_panic: bool,
-    ) -> Option<GlobalGuard<'_, T>> {
-        let init_fn = match &mut *mutex_guard {
-            InitState::Initialized(_) => {
-                // SAFETY: `mutex_guard` is `Initialized`.
-                return Some(unsafe { GlobalGuard::new_unchecked(mutex_guard) });
-            }
-            InitState::Failed => {
-                return None;
-            }
-            InitState::Pending(init_fn) => init_fn,
-        };
-
-        // Unwinding should be safe here, as there is no unsafe code relying on it.
-        let init_fn = std::panic::AssertUnwindSafe(init_fn);
-        match std::panic::catch_unwind(init_fn) {
-            Ok(value) => *mutex_guard = InitState::Initialized(value),
-            Err(e) => {
-                eprintln!("panic during Global<T> initialization");
-                *mutex_guard = InitState::Failed;
-
-                if may_panic {
-                    std::panic::resume_unwind(e);
-                } else {
-                    // Note: this currently swallows panic.
-                    return None;
+        let guard = self.value.try_lock().map_err(|x| match x {
+            TryLockError::Poisoned(x) => {
+                let g = x.into_inner();
+                match std::panic::catch_unwind(|| g.get_or_init(self.init_fn)) {
+                    Ok(_) => GlobalLockError::Poisoned {
+                        // SAFETY: see get_or_init.
+                        circumvent: unsafe { GlobalGuard::new_unchecked(g) },
+                    },
+                    Err(_) => GlobalLockError::InitFailed,
                 }
             }
-        };
-
-        // SAFETY: `mutex_guard` was either set to `Initialized` above, or we returned from the function.
-        Some(unsafe { GlobalGuard::new_unchecked(mutex_guard) })
+            TryLockError::WouldBlock => GlobalLockError::WouldBlock,
+        })?;
+        std::panic::catch_unwind(|| guard.get_or_init(self.init_fn))
+            .map_err(|_| GlobalLockError::InitFailed)?;
+        Ok(unsafe { GlobalGuard::new_unchecked(guard) })
```

Comment:
It's not obvious why this is called again here, could you elaborate?

Also, `unsafe` would need `// SAFETY`, can be kept short like above.

Reply:
i hope ive clarified it now